clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -328,6 +331,73 @@ check_kernel_xfer()
         if (!run_checker(memrefs, false))
             return false;
     }
+    // Multiple nested signals without any intervening instr between the nested
+    // signals.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(1, 1),
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 2),
+            gen_instr(1, 101),
+            // First nested signal.
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 102),
+            gen_instr(1, 201),
+            // XXX: This marker value is actually not guaranteed, yet the checker
+            // requires it and the view tool prints it.
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_XFER, 202),
+            // No intervening instr here. Should use instr at pc = 101 for
+            // pre-signal instr check on return.
+            // Second nested signal.
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 102),
+            gen_instr(1, 201),
+            // XXX: This marker value is actually not guaranteed, yet the checker
+            // requires it and the view tool prints it.
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_XFER, 202),
+            gen_instr(1, 102),
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_XFER, 103),
+            gen_instr(1, 2),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Trace starts in a signal.
+    {
+        std::vector<memref_t> memrefs = {
+            // Already inside the first signal.
+            gen_instr(1, 11),
+            // XXX: This marker value is actually not guaranteed, yet the checker
+            // requires it and the view tool prints it.
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_XFER, 12),
+            // Should skip the pre-signal instr check and the kernel_event marker
+            // equality check, since we did not see the beginning of the signal in
+            // the trace.
+            gen_instr(1, 2),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Trace starts in a signal with a back-to-back signal without any intervening
+    // instr after we return from the first one.
+    {
+        std::vector<memref_t> memrefs = {
+            // Already inside the first signal.
+            gen_instr(1, 11),
+            // XXX: This marker value is actually not guaranteed, yet the checker
+            // requires it and the view tool prints it.
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_XFER, 12),
+            // No intervening instr here. Should skip pre-signal instr check on
+            // return; this is a special case as it would require *removing* the
+            // pc = 1 instr from pre_signal_instr_ as it was not in this newly
```

Comment:
?  There is no pc=1 instr here: you mean pc=11?

Reply:
Ah yes. It should be pc = 11.

Reply:
Done.